File: haskell/014.hs
collatz :: Integral a => a -> a
collatz n
    | n == 1 = 1
    | even n = 1 + collatz(div n 2)
    |  odd n = 1 + collatz(3*n+1)

chains = [(num, collatz num) | num <- [1..1000000]]
highestChain = maximum ([chain | (num, chain) <- chains])
solution = take 1 [num | (num, chain) <- chains, chain == highestChain]

main = do print solution

File: c/014.c
#include <stdio.h>
#include <stdint.h>
#include "Euler.h"

int main(void)
{
    uint64_t num;
    int chain;
    int max_starting = 1;
    int max = 0;
    for (num = 1; num < 1e6; num++) {
        chain = collatz(num);
        if (chain > max) {
            max_starting = num;
            max = chain;
        }
    }
    printf("%d\n", max_starting);
    return 0;
}
